sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SyntaxTrackers/CSharpObjectInitializationTracker.cs

```diff
@@ -212,8 +212,11 @@ private bool IsLaterAssignedWithAllowedValue(ObjectCreationExpressionSyntax obje
 
             var variableSymbol = GetAssignedVariableSymbol(objectCreation, semanticModel);
 
+            var nextStatements = GetNextStatements(statement);
+            var innerStatements = GetInnerStatements(statement);
+
             return variableSymbol != null
-                && GetNextStatements(statement)
+                && nextStatements.Union(innerStatements)
                     .OfType<ExpressionStatementSyntax>()
                     .Select(x => x.Expression)
                     .OfType<AssignmentExpressionSyntax>()
```

Comment:
If I'm not missing something we can simplify the code a bit by replacing `GetNextStatements` and `GetInnerStatements` with:
```
private static IEnumerable<AssignmentExpressionSyntax> GetAssignmentExpressions(StatementSyntax statement) =>
    statement.Parent.DescendantNodes().OfType<AssignmentExpressionSyntax>();
```

then we could replace the return with: 
`return variableSymbol != null && GetAssignmentExpressions(statement).Any(TrackedPropertySetWithAllowedValue);`

Reply:
your approach changes the functionality (because it takes all the nodes, not only the ones in the same block and the "inner block" of the using/if in the samples)

for example, we currently have tests 
```
            var reader = new XmlTextReader(Url); // Noncompliant
            if (foo)
            {
                reader.ProhibitDtd = true; // this is set conditionally, so not enough
            }

```

and this will not raise anymore (the idea is that after the if, the `reader` object is still vulnerable)

also, there are failing tests 

```
            var c2 = new HttpCookie("c"); // Noncompliant, HttpOnly is set conditionally
            if (foo)
            {
                c2.HttpOnly = true;
            }

```

and I don't want to change the functionality of the security hotspots that are using this base class